nupic/embodied/multitask/algos/custom_mtsac.py

```diff
@@ -196,6 +206,226 @@ def train(self, trainer):
+                "The number of tasks in the environment does "
+                "not match self._num_tasks. Are you sure that you passed "
+                "The correct number of tasks?")
+        
+        with autocast(enabled=self._fp16):
+            return torch.mm(one_hots, log_alpha.unsqueeze(0).t()).squeeze()
+
+    def _temperature_objective(self, log_pi, samples_data):
+        """Compute the temperature/alpha coefficient loss.
+        Args:
+            log_pi(torch.Tensor): log probability of actions that are sampled
+                from the replay buffer. Shape is (1, buffer_batch_size).
+            samples_data (dict): Transitions(S,A,R,S') that are sampled from
+                the replay buffer. It should have the keys 'observation',
+                'action', 'reward', 'terminal', and 'next_observations'.
+        Note:
+            samples_data's entries should be torch.Tensor's with the following
+            shapes:
+                observation: :math:`(N, O^*)`
+                action: :math:`(N, A^*)`
+                reward: :math:`(N, 1)`
+                terminal: :math:`(N, 1)`
+                next_observation: :math:`(N, O^*)`
+        Returns:
+            torch.Tensor: the temperature/alpha coefficient loss.
+        """
+        alpha_loss = 0
+
+        with autocast(enabled=self._fp16):
+            if self._use_automatic_entropy_tuning:
+                alpha_loss = (-(self._get_log_alpha(samples_data)) *
+                            (log_pi.detach() + self._target_entropy)).mean()
+
+            return alpha_loss
+
+    def _actor_objective(self, samples_data, new_actions, log_pi_new_actions):
+        """Compute the Policy/Actor loss.
+        Args:
+            samples_data (dict): Transitions(S,A,R,S') that are sampled from
+                the replay buffer. It should have the keys 'observation',
+                'action', 'reward', 'terminal', and 'next_observations'.
+            new_actions (torch.Tensor): Actions resampled from the policy based
+                based on the Observations, obs, which were sampled from the
+                replay buffer. Shape is (action_dim, buffer_batch_size).
+            log_pi_new_actions (torch.Tensor): Log probability of the new
+                actions on the TanhNormal distributions that they were sampled
+                from. Shape is (1, buffer_batch_size).
+        Note:
+            samples_data's entries should be torch.Tensor's with the following
+            shapes:
+                observation: :math:`(N, O^*)`
+                action: :math:`(N, A^*)`
+                reward: :math:`(N, 1)`
+                terminal: :math:`(N, 1)`
+                next_observation: :math:`(N, O^*)`
+        Returns:
+            torch.Tensor: loss from the Policy/Actor.
+        """
+        obs = samples_data["observation"]
+
+        with torch.no_grad():
+            alpha = self._get_log_alpha(samples_data).exp()
+
+        with autocast(enabled=self._fp16):
```

Comment:
Nit: if you can have a hierarchy of contexts, wouldn't be easier to just put autocast prior to line 299 instead of having it here and inside get log alpha?

Reply:
Yep I can do this

Reply:
Actually, I think it might be cleaner this way - `_get_log_alpha()` already has covered by a autocast context, so it's redundant to do it again. 

Every method should automatically require its components to be covered by an autocast context if FP16 is used, instead of autocasting in some main method where the other functions are called. If I changed it here, I would have to change it for  `_critic_objective` and `_actor_objective`, and it might look messy.

What do you think?

Reply:
Got it. Yeah, agree with you.